mesosphere/charts
# velero-minio: fix the data loss issue

Velero chart 3.0.0 and 3.0.1 have data loss issue. The reason is because
we didn't set the `mountPath` for PV properly in minio. The default
`mountPath` is `/export` while the `bucketRoot` is at `/data`. This
cause velero-minio to start a fresh disk from the container imagefs,
rather than from the PV.

However, after fixing the `mountPath`, we noticed that the minio server
won't start due to "invalid credentials". After researching a bit, we
found that minio requires the original access secret to decrypt the data
on the PV. As a result, we cannot use a delete and re-install policy
because the minio credential secret (namely `minio-creds-secret) created
by the init container has an owner reference to the velero deployment.

Previously, we annotate the Addon to use a delete and re-install policy
to avoid the helm upgrade error because the statefulset `minio` already
exists (created by the minio operator in ealier 2.x.y versions, thus is
not tracked by helm). And we have to use the same statefulset name
beacuse PVC name is derived from the statefulset name.

To solve the issue, we have to make the helm upgrade work. Thus in this
patch, we install a `pre-install` job to delete objects created by the
minio operator in earlier 2.x.y versions, including the statefulset
`minio`. For future upgrade (upgrade from helm based minio), this
cleanup job should be a no-op.

Manually tested this by doing the following:
1. `konvoy up` using 1.4.1 (KBA `stable-1.16-1.2.0`)

2. Observed that a backup is automatically created:
```bash
➜  velero backup get 
NAME                                       STATUS      CREATED                         EXPIRES   STORAGE LOCATION   SELECTOR
velero-kubeaddons-default-20200327032731   Completed   2020-03-26 20:27:31 -0700 PDT   29d       default            <none>
```

3.  `helm upgrade --install velero-kubeaddons -f myvalues.yaml --namespace velero .` and observed that the upgrade is successful. Check that the previous backup is still there
```bash
➜  velero backup get  
NAME                                       STATUS      CREATED                         EXPIRES   STORAGE LOCATION   SELECTOR
velero-kubeaddons-default-20200327032731   Completed   2020-03-26 20:27:31 -0700 PDT   29d       default            <none>
```
Also checking the filesystem inside minio pod, verified that the bucketRoot is correct:
```bash
➜  k exec -ti -n velero minio-0  /bin/sh                                         
/ # df -Th
Filesystem           Type            Size      Used Available Use% Mounted on
overlay              overlay       159.9G      4.3G    155.6G   3% /
tmpfs                tmpfs          64.0M         0     64.0M   0% /dev
tmpfs                tmpfs          15.4G         0     15.4G   0% /sys/fs/cgroup
/dev/nvme2n1         ext4            9.7G     36.8M      9.7G   0% /data
/dev/nvme0n1p1       xfs            80.0G      1.6G     78.4G   2% /etc/hosts
/dev/nvme0n1p1       xfs            80.0G      1.6G     78.4G   2% /dev/termination-log
/dev/nvme1n1         xfs           159.9G      4.3G    155.6G   3% /etc/hostname
/dev/nvme1n1         xfs           159.9G      4.3G    155.6G   3% /etc/resolv.conf
shm                  tmpfs          64.0M         0     64.0M   0% /dev/shm
tmpfs                tmpfs          15.4G     12.0K     15.4G   0% /run/secrets/kubernetes.io/serviceaccount
tmpfs                tmpfs          15.4G         0     15.4G   0% /proc/acpi
tmpfs                tmpfs          64.0M         0     64.0M   0% /proc/kcore
tmpfs                tmpfs          64.0M         0     64.0M   0% /proc/keys
tmpfs                tmpfs          64.0M         0     64.0M   0% /proc/timer_list
tmpfs                tmpfs          64.0M         0     64.0M   0% /proc/timer_stats
tmpfs                tmpfs          64.0M         0     64.0M   0% /proc/sched_debug
tmpfs                tmpfs          15.4G         0     15.4G   0% /proc/scsi
tmpfs                tmpfs          15.4G         0     15.4G   0% /sys/firmware
```

4. Create another backup
```bash
➜  velero backup create jie
Backup request "jie" submitted successfully.
Run `velero backup describe jie` or `velero backup logs jie` for more details.
```
```
➜  velero backup get
NAME                                       STATUS      CREATED                         EXPIRES   STORAGE LOCATION   SELECTOR
jie                                        Completed   2020-03-26 20:37:08 -0700 PDT   29d       default            <none>
velero-kubeaddons-default-20200327032731   Completed   2020-03-26 20:27:31 -0700 PDT   29d       default            <none>
```

5.  Try upgrade again with the same chart to test idempotency: `helm upgrade --install velero-kubeaddons -f myvalues.yaml --namespace velero .` Observed that the upgrade succeeds.

6. Verified that all previous backups are there:
```bash
➜  velero backup get                      
NAME                                       STATUS      CREATED                         EXPIRES   STORAGE LOCATION   SELECTOR
jie                                        Completed   2020-03-26 20:37:08 -0700 PDT   29d       default            <none>
velero-kubeaddons-default-20200327032731   Completed   2020-03-26 20:27:31 -0700 PDT   29d       default            <none>
```

`myvalues.yaml` is like the following:
```yaml
configuration:
  provider: "aws"
  backupStorageLocation:
    name: "aws"
    bucket: "velero"
    config:
      region: "fallback"     # enables non-production fallback minio backend
      s3ForcePathStyle: true # allows usage of fallback backend
      s3Url: http://minio.velero.svc:9000
  volumeSnapshotLocation:
    name: "aws"
    config:
      region: "fallback"
credentials:
  secretContents:
    cloud: "placeholder"
schedules:
  default:
    schedule: "0 0 * * *"
metrics:
  enabled: true
  service:
    labels:
      servicemonitor.kubeaddons.mesosphere.io/path: "metrics"
initContainers:
- name: initialize-velero
  image: mesosphere/kubeaddons-addon-initializer:v0.2.5
  args: ["velero"]
  env:
  - name: "VELERO_MINIO_FALLBACK_SECRET_NAME"
    value: "velero-kubeaddons"
minioBackend: true
minio:
  mode: distributed
  defaultBucket:
    enabled: true
    name: velero
  bucketRoot: "/data"
  mountPath: "/data"
  existingSecret: minio-creds-secret
  livenessProbe:
    initialDelaySeconds: 120
    periodSeconds: 20
  resources:
    requests:
      memory: 256Mi
      cpu: 250m
    limits:
      memory: 512Mi
      cpu: 750m
  persistence:
    volumeTemplatePrefix: data
  statefulSetNameOverride: minio
  ingress:
    enabled: true
    hosts:
    - ""
    annotations:
      kubernetes.io/ingress.class: traefik
      traefik.ingress.kubernetes.io/frontend-entry-points: velero-minio
```
